Modules/_testexternalinspection.c

```diff
@@ -433,44 +469,623 @@ read_string(pid_t pid, _Py_DebugOffsets* debug_offsets, void* address, char* buf
     return 0;
 }
 
-void*
-get_py_runtime(pid_t pid)
+
+static inline int
+read_ptr(pid_t pid, uintptr_t address, uintptr_t *ptr_addr)
 {
-#if defined(__linux__)
-    return get_py_runtime_linux(pid);
-#elif defined(__APPLE__) && TARGET_OS_OSX
-    return get_py_runtime_macos(pid);
-#else
-    return NULL;
-#endif
+    int bytes_read = read_memory(pid, address, sizeof(void*), ptr_addr);
+    if (bytes_read == -1) {
+        return -1;
+    }
+    return 0;
+}
+
+static inline int
+read_ssize_t(pid_t pid, uintptr_t address, Py_ssize_t *size)
+{
+    int bytes_read = read_memory(pid, address, sizeof(Py_ssize_t), size);
+    if (bytes_read == -1) {
+        return -1;
+    }
+    return 0;
 }
 
 static int
-parse_code_object(
-        int pid,
-        PyObject* result,
-        struct _Py_DebugOffsets* offsets,
-        void* address,
-        void** previous_frame)
+read_py_ptr(pid_t pid, uintptr_t address, uintptr_t *ptr_addr)
 {
-    void* address_of_function_name;
-    read_memory(
+    if (read_ptr(pid, address, ptr_addr)) {
+        return -1;
+    }
+    *ptr_addr &= ~Py_TAG_BITS;
+    return 0;
+}
+
+static int
+read_char(pid_t pid, uintptr_t address, char *result)
+{
+    int bytes_read = read_memory(pid, address, sizeof(char), result);
+    if (bytes_read < 0) {
+        return -1;
+    }
+    return 0;
+}
+
+static int
+read_int(pid_t pid, uintptr_t address, int *result)
+{
+    int bytes_read = read_memory(pid, address, sizeof(int), result);
+    if (bytes_read < 0) {
+        return -1;
+    }
+    return 0;
+}
+
+static int
+read_pyobj(pid_t pid, uintptr_t address, PyObject *ptr_addr)
+{
+    int bytes_read = read_memory(pid, address, sizeof(PyObject), ptr_addr);
+    if (bytes_read < 0) {
+        return -1;
+    }
+    return 0;
+}
+
+static PyObject *
+read_py_str(
+    pid_t pid,
+    _Py_DebugOffsets* debug_offsets,
+    uintptr_t address,
+    ssize_t max_len
+) {
+    assert(max_len > 0);
+
+    PyObject *result = NULL;
+
+    char *buf = (char *)PyMem_RawMalloc(max_len);
+    if (buf == NULL) {
+        PyErr_NoMemory();
+        return NULL;
+    }
+    if (read_string(pid, debug_offsets, address, buf, max_len)) {
+        goto err;
+    }
+
+    result = PyUnicode_FromString(buf);
+    if (result == NULL) {
+        goto err;
+    }
+
+    PyMem_RawFree(buf);
+    assert(result != NULL);
+    return result;
+
+err:
+    PyMem_RawFree(buf);
+    return NULL;
+}
+
+static long
+read_py_long(
+    pid_t pid,
+    _Py_DebugOffsets* offsets,
+    uintptr_t address) {
+    unsigned int shift = PYLONG_BITS_IN_DIGIT;
+
+    ssize_t size;
+    uintptr_t lv_tag;
+    int bytes_read = read_memory(pid, address + offsets->long_object.lv_tag, sizeof(uintptr_t), &lv_tag);
```

Comment:
You might want to reflow it under 80 chars.

Reply:
OK, I reformatted all of the test code to be under 80.